File: design_pattern/state.ts
// ============================================================
// State Pattern — 对象根据内部状态改变行为
// 对比 Rust: 09_state.rs
// 运行: npx ts-node 09_state.ts
// ============================================================

type OrderStatus = "Pending" | "Paid" | "Shipped" | "Delivered" | "Cancelled";

class Order {
  private status: OrderStatus = "Pending";

  constructor(public readonly id: number, public readonly item: string) {}

  pay(): void {
    if (this.status !== "Pending") throw new Error(`状态 ${this.status} 不允许支付`);
    this.status = "Paid";
    console.log(`[Order#${this.id}] 支付成功 -> ${this.status}`);
  }

  ship(): void {
    if (this.status !== "Paid") throw new Error(`状态 ${this.status} 不允许发货`);
    this.status = "Shipped";
    console.log(`[Order#${this.id}] 已发货 -> ${this.status}`);
  }

  deliver(): void {
    if (this.status !== "Shipped") throw new Error(`状态 ${this.status} 不允许确认收货`);
    this.status = "Delivered";
    console.log(`[Order#${this.id}] 已送达 -> ${this.status}`);
  }

  cancel(): void {
    if (this.status === "Delivered" || this.status === "Cancelled")
      throw new Error(`状态 ${this.status} 不允许取消`);
    this.status = "Cancelled";
    console.log(`[Order#${this.id}] 已取消 -> ${this.status}`);
  }

  getStatus(): OrderStatus { return this.status; }
}

const tryOp = (fn: () => void) => {
  try { fn(); } catch (e: any) { console.log("  [错误]", e.message); }
};

// --- main ---
console.log("=== State Pattern ===\n");

console.log("--- 正常流程 ---");
const order = new Order(1001, "MacBook Pro");
console.log("初始:", order.getStatus());
tryOp(() => order.pay());
tryOp(() => order.ship());
tryOp(() => order.deliver());

console.log("\n--- 非法转换 ---");
const order2 = new Order(1002, "iPhone");
tryOp(() => order2.ship());
tryOp(() => order2.pay());
tryOp(() => order2.pay());
tryOp(() => order2.cancel());
tryOp(() => order2.cancel());

// Rust 关键差异：
// - Rust 的 match 是穷举的，漏掉状态编译报错
// - TS 的 if/else 漏掉状态只是运行时 bug

File: design_pattern/state.rs
// ============================================================
// State Pattern — 对象根据内部状态改变行为，状态转换显式化
// 对比 TS: 09_state.ts
// 运行: cargo run --bin state
// ============================================================

#[derive(Debug, Clone, PartialEq)]
enum OrderStatus {
    Pending,
    Paid,
    Shipped,
    Delivered,
    Cancelled,
}

struct Order {
    id: u64,
    #[allow(dead_code)]
    item: String,
    status: OrderStatus,
}

impl Order {
    fn new(id: u64, item: &str) -> Self {
        Self { id, item: item.into(), status: OrderStatus::Pending }
    }

    fn pay(&mut self) -> Result<(), String> {
        match self.status {
            OrderStatus::Pending => {
                self.status = OrderStatus::Paid;
                println!("[Order#{}] 支付成功 -> {:?}", self.id, self.status);
                Ok(())
            }
            _ => Err(format!("状态 {:?} 不允许支付", self.status)),
        }
    }

    fn ship(&mut self) -> Result<(), String> {
        match self.status {
            OrderStatus::Paid => {
                self.status = OrderStatus::Shipped;
                println!("[Order#{}] 已发货 -> {:?}", self.id, self.status);
                Ok(())
            }
            _ => Err(format!("状态 {:?} 不允许发货", self.status)),
        }
    }

    fn deliver(&mut self) -> Result<(), String> {
        match self.status {
            OrderStatus::Shipped => {
                self.status = OrderStatus::Delivered;
                println!("[Order#{}] 已送达 -> {:?}", self.id, self.status);
                Ok(())
            }
            _ => Err(format!("状态 {:?} 不允许确认收货", self.status)),
        }
    }

    fn cancel(&mut self) -> Result<(), String> {
        match self.status {
            OrderStatus::Delivered | OrderStatus::Cancelled => {
                Err(format!("状态 {:?} 不允许取消", self.status))
            }
            _ => {
                self.status = OrderStatus::Cancelled;
                println!("[Order#{}] 已取消 -> {:?}", self.id, self.status);
                Ok(())
            }
        }
    }
}

fn try_op(result: Result<(), String>) {
    if let Err(e) = result {
        println!("  [错误] {}", e);
    }
}

fn main() {
    println!("=== State Pattern ===\n");

    println!("--- 正常流程 ---");
    let mut order = Order::new(1001, "MacBook Pro");
    println!("初始: {:?}", order.status);
    try_op(order.pay());
    try_op(order.ship());
    try_op(order.deliver());

    println!("\n--- 非法转换 ---");
    let mut order2 = Order::new(1002, "iPhone");
    try_op(order2.ship());    // Pending 不能直接发货
    try_op(order2.pay());
    try_op(order2.pay());     // 不能重复支付
    try_op(order2.cancel());
    try_op(order2.cancel()); // 已取消不能再取消
}

// Rust 关键差异：
// - enum + match 表达状态机是 Rust 最惯用的写法
// - match 是穷举的，忘处理某个状态会编译报错
// - Result<(), String> 强制调用方处理错误
